NE-1325: External DNS Operator support for Shared VPCs (#1436)
openshift/enhancements: enhancements/installer/aws-cross-account-dns-zone.md

Reviewer: Did you address anything that would necessitate removing the comments about the open questions?  If not, just leave them there.

Author: I didn't address anything, but the lines needed to be cleaned up as they don't make sense anymore.

For `The name of the field is discussed further in open questions.` and `Please see Open Questions for further discussion of the install config.`, there are no open questions, so these didn't make sense.

CC: @patrickdillon was the original author

Reviewer: I'll let him review this change then.  I'll just review the stuff in the new EP.

Author: I'll ping him for a review.

Reviewer: This EP specifies a target of 4.14.  If the enhancement is not targetting 4.14, it's not quite right to mention it within this EP.  Why not update the ExternalDNS EP, and point to the update from this EP?

Author: Hmm. Good question. The Enhanced DNS work won't target any OCM release, since it's an along side item. I think we could make that clear in this enhancement to address your concern.

But I don't know if it's better to update the External DNS Operator EP. I'm a bit ambivalent. I've [OK-ed](https://redhat-internal.slack.com/archives/C054PJX8LE8/p1687807598146529) updating this enhancement with the PM of the Shared VPC feature and they seemed onboard, but that's not to say it's not the right place.

This External DNS update work is a part of the same [Shared VPC Feature](https://issues.redhat.com/browse/OCPSTRAT-377). Should organize per feature vs. per operator/project/repo (e.g. external-dns-operator)?

I had actually asked @Miciah this in the beginning, as I wasn't sure where to start, and he suggested this adding to this enhancement. Maybe he could also shine some light on these question.

Author: Update: I updated to move to it's own enhancement. After the Shared VPC features split into separate features (https://issues.redhat.com/browse/OCPSTRAT-730), it felt even more decoupled. But regardless of that, it started to feel shoehorned into this enhancement like I think you were implying.

I didn't feel like the existing External DNS enhancement was the right place, as that was outlining the design of bringing ExternalDNS Operator to OpenShift, and now we are much past that. So i went with a new one. Let me know your thoughts.

    hostedZone: Z00147933I3NWOQ6M4699
    hostedZoneRole: arn:aws:iam::<account-a>:role/<role-name>
```

### API: External DNS Operator

This enhancement also extends [External DNS Operator](https://github.com/openshift/external-dns-operator), an add-on

> operator deployable using Operator Lifecycle Manager (OLM), to support cross account DNS zones in AWS. We extend the
> `ExternalDNS` API object by adding the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).
> Refer to [externaldns_types.go](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go)

Reviewer 2: Tiny copyedits:
```suggestion
operator deployable using Operator Lifecycle Manager (OLM), to support cross-account DNS zones in AWS. We extend the
`ExternalDNS` API object by adding the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).
Refer to [`externaldns_types.go`](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go)
```
I was going to suggest using a permalink, but in this case (pointing the the v1beta1 API, without any line number), I suppose it's appropriate to use the main branch reference.  

Consider adding a note that the proposed API change is purely additive and is therefore a non-breaking, backwards-compatible change.

    hostedZone: Z00147933I3NWOQ6M4699
    hostedZoneRole: arn:aws:iam::<account-a>:role/<role-name>
```

### API: External DNS Operator

This enhancement also extends [External DNS Operator](https://github.com/openshift/external-dns-operator), an add-on
operator deployable using Operator Lifecycle Manager (OLM), to support cross-account DNS zones in AWS. We extend the
`ExternalDNS` API object by adding the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).
Refer to [`externaldns_types.go`](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go)
for the existing API structure of the `ExternalDNS` object. These API updates are purely additive and, therefore,
represent a non-breaking, backwards-compatible change.

```go
type ExternalDNSAWSProviderOptions struct {
     [...]
     // RoleARN contains the ARN of a IAM role that will be assumed when using the AWS API.
     // It provides the ability to use a hosted zone in another AWS account.
     //
     // +kubebuilder:validation:Optional

>      // +optional

Author: Note: I was going to add validation for the roleARN here.

Author: Done, I forgot to close this comment

    hostedZone: Z00147933I3NWOQ6M4699
    hostedZoneRole: arn:aws:iam::<account-a>:role/<role-name>
```

### API: External DNS Operator

> This enhancement also extends [External DNS Operator](https://github.com/openshift/external-dns-operator), an add-on
> operator deployable using Operator Lifecycle Manager (OLM), to support cross-account DNS zones in AWS. We extend the
> `ExternalDNS` API object by adding the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).

Reviewer: ```suggestion
[ExternalDNS Operator](https://github.com/openshift/external-dns-operator) allows you to deploy and manage [ExternalDNS](https://github.com/kubernetes-sigs/external-dns), a cluster-internal component which makes Kubernetes resources discoverable through public DNS servers.  This enhancement extends ExternalDNS Operator to support cross-account DNS zones in AWS. We extend the`ExternalDNS` API object by adding the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).
```

Author: Done.

    hostedZone: Z00147933I3NWOQ6M4699
    hostedZoneRole: arn:aws:iam::<account-a>:role/<role-name>
```

### API: ExternalDNS Operator

[ExternalDNS Operator](https://github.com/openshift/external-dns-operator) allows you to deploy and manage [ExternalDNS](https://github.com/kubernetes-sigs/external-dns),
a cluster-internal component which makes Kubernetes resources discoverable through public DNS servers. This enhancement
extends the ExternalDNS Operator to support cross-account DNS zones in AWS. We extend the`ExternalDNS` API object by adding
the field `RoleARN` to `ExternalDNSAWSProviderOptions` struct (`spec.provider.aws`).
Refer to [`externaldns_types.go`](https://github.com/openshift/external-dns-operator/blob/main/api/v1beta1/externaldns_types.go)
for the existing API structure of the `ExternalDNS` object.

> In addition to adding `RoleARN`, the `Credentials` field was updated to be optional. Prior to this change, the `Credentials`

Reviewer 3: >  the `Credentials` field was updated to be optional

I think we have to consult the API team on that change. It's a tricky thing, there are a lot of caveats: how the object is stored in etcd, API calls vs manifests, server-side apply, etc. My fear is that even a webhook may be not enough here.

Author: I don't think I'm following completely. Could you elaborate on your concern here? What scenarios are you envisioning `optional` being an issue? 

Is your concern with just `Credentials` being optional? Or is your concern with the updating `Credentials` from `required` to `optional`?

Reviewer 3: > Or is your concern with the updating Credentials from required to optional?

Yes.

I think I came up with an example which may cause issues if we don't bump a new ExternalDNS api version.
Imagine a client application calling ExternalDNS API:
```go
func getAWSCredentialsName() string { 
  extdns := &extdnsv1beta1.ExternalDNS{}
  ocClient.Get(extdns)
  if extdns.spec.provider.type == "AWS" && extdns.spec.provider.aws != nil {
    return extdns.spec.provider.aws.Credentials
  }
  return ""
}
```
This code will become faulty.

Author: Ah okay I see. The assumption they were making was that `extdns.Spec.Provider.AWS.Credentials` was always not-nil because it was required.

So sound like we need to bump the version of External DNS operator API. Looking back at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility, we violate `which fields are required and which are not`.

So...how do we go about bumping the API? Can we do a `v1beta2` similar to what Gateway API has [done](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1alpha2)?

Reviewer 3: > So sound like we need to bump the version of External DNS operator API. Looking back at https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api_changes.md#on-compatibility, we violate which fields are required and which are not.

Yes.

> Can we do a v1beta2 similar to what Gateway API has [done](https://gateway-api.sigs.k8s.io/references/spec/#gateway.networking.k8s.io%2fv1alpha2)?

Yeah, I think `v1beta2` should be fine.

> So...how do we go about bumping the API?

It should be a new api version with a conversion webhook to handle a situation when `ExternalDNSAWSProviderOptions` is empty. The conversion webhook would force us to change the install mode for the operator to `AllNamespaces`. I did a similar exercise when `v1` version of ALBO was released ([PR](https://github.com/openshift/aws-load-balancer-operator/pull/93)). 

Overall it's a pity we have to do all this just to a add a new API field. We didn't think about setting the field optional from [the introduction of the API](https://github.com/openshift/external-dns-operator/blame/main/api/v1alpha1/externaldns_types.go#L258) and now it strikes back.

Maybe API team may come up with some tricks of how to approach this differently (without a new API version). For instance, a new field or a new union, I gave myself 30 mins to think of it but didn't come with any decent idea.

Author: Resolving thread as we discussed a way forward in https://redhat-internal.slack.com/archives/CE4L0F143/p1691007201050019?thread_ts=1690552114.405509&cid=CE4L0F143

For more context, decision is documented in https://github.com/openshift/external-dns-operator/pull/195/

field was required if `spec.provider.aws` was specified. However, OpenShift clusters are exempt from explicitly
providing `Credentials` since the ExternalDNS Operator generates a `CredentialRequest` automatically. Consequently, to
enable OpenShift clusters to specify `RoleARN` without `Credentials`, it is necessary to make the `Credentials` field
optional. The validation webhook takes care of guaranteeing that non-OpenShift clusters include credentials through the

> API.

Reviewer 3: So on OpenShift you can have empty `aws` field then:
```yaml
provider:
  type: aws
  aws: {}
```

Author: Yes I suppose, but `omitempty` is [specified](https://github.com/openshift/external-dns-operator/blob/68bdec5a19686d0ccdb3ed1f657dbfd14809b338/api/v1beta1/externaldns_types.go#L215) on the `aws` field.

Is this problematic?

Reviewer 3: I will not pass the validating webhook due to [the validate credentials logic for AWS](https://github.com/openshift/external-dns-operator/blob/68bdec5a19686d0ccdb3ed1f657dbfd14809b338/api/v1beta1/externaldns_webhook.go#L138-L140).

```

#### External DNS Operator Details

Prior to this enhancement, the [External DNS Operator API](https://github.com/openshift/external-dns-operator/blob/main/api)
did not allow users to configure an AWS role ARN field. As a result, there wasn't a supported way to use the External
DNS Operator to create DNS records in another AWS account within a shared VPC.

The External DNS Operator configures and runs the [External DNS](https://github.com/openshift/external-dns) binary.

> Conveniently, External DNS already supports the [--aws-assume-role](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/docs/faq.md#can-external-dns-manageaddremove-records-in-a-hosted-zone-which-is-setup-in-different-aws-account)

Reviewer 2: Tiny copyedit:
```suggestion
Conveniently, External DNS already supports the [`--aws-assume-role`](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/docs/faq.md#can-external-dns-manageaddremove-records-in-a-hosted-zone-which-is-setup-in-different-aws-account)
```

argument, which uses the specified AWS role ARN when creating new DNS records. Therefore, to support shared VPCs with
the External DNS Operator, we update the [API](#API-External-DNS-Operator) to expose the `--aws-assume-role` argument
for the External DNS binary.

> By design, External DNS can update both public and private DNS zones. Therefore, External DNS Operator users are able
> to use both public and private DNS zones in other accounts, provided the role ARN has appropriate permissions.

Reviewer 2: I don't think this statement logically follows without this addition:
```suggestion
By design, External DNS can update both public and private DNS zones using credentials that the cluster-admin provides. Therefore, External DNS Operator users are able
to use both public and private DNS zones in other accounts, provided the role ARN has appropriate permissions.
```

Author: Sure, provides clarity.

More details on using a role ARN to create DNS records in another AWS account will be documented in External DNS

> Operator's [Docs](https://github.com/openshift/external-dns/tree/master/docs).

Reviewer 2: ```suggestion
Operator's [Docs](https://github.com/openshift/external-dns-operator/tree/main/docs).
```

Author: Whoops, good catch.

```

#### External DNS Operator Details

Prior to this enhancement, the [External DNS Operator API](https://github.com/openshift/external-dns-operator/blob/main/api)
did not allow users to configure an AWS role ARN field. As a result, there wasn't a supported way to use the External

> DNS Operator to create DNS records in another AWS account within a shared VPC.

Reviewer: Is there any built-in installer support ?  I thought ExternalDNS was a day-two installation option, meaning none of it is handled by the OpenShift installer.  How does this new functionality integrate with OpenShift Installer?

Author: You are right, ExternalDNS is a day-two operation, installed as a addon operator.

It doesn't integrate with the installer for this feature. It effectively provides the same feature as the installer and Ingress Operator implement together: support for creating a DNS record in an another zone.

Conveniently, External DNS already supports the [`--aws-assume-role`](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/docs/faq.md#can-external-dns-manageaddremove-records-in-a-hosted-zone-which-is-setup-in-different-aws-account)
argument, which uses the specified AWS role ARN when creating new DNS records. Therefore, to support shared VPCs with
the External DNS Operator, we update the [API](#API-External-DNS-Operator) to expose the `--aws-assume-role` argument
for the External DNS binary.

By design, External DNS can update both public and private DNS zones using credentials that the cluster-admin provides.
Therefore, External DNS Operator users are able to use the new [API](#API-External-DNS-Operator) to update both public
and private DNS zones in other accounts, provided the role ARN has appropriate permissions.

More details on using a role ARN to create DNS records in another AWS account will be documented in External DNS
Operator's [Docs](https://github.com/openshift/external-dns-operator/tree/main/docs).

Note that External DNS Operator updates are asynchronous to OpenShift's standard release of OpenShift Container

> Platform. Therefore, any updates to External DNS Operator are considered out of payload.

Reviewer: Consider mentioning this in Risks and Mitigations or Version Skew, or both.

Author: Good idea, I added it to Version Skew, as that seemed appropriate. I'm not sure if it's a risk or needs a mitigation, I documented that it isn't a risk in Version Skew.

### Risks and Mitigations

> Add-on operators needing permissions will not work unless updated to use the role ARN.

Author: Done.

```

#### ExternalDNS Operator Details

Prior to this enhancement, the [ExternalDNS Operator API](https://github.com/openshift/external-dns-operator/blob/main/api)
did not allow users to configure an AWS role ARN field. As a result, there wasn't a supported way to use the External
DNS Operator to create DNS records in another AWS account within a shared VPC.

The ExternalDNS Operator configures and runs the [ExternalDNS](https://github.com/openshift/external-dns) binary.

> Conveniently, ExternalDNS already supports the [`--aws-assume-role`](https://github.com/openshift/external-dns/blob/fe00b4b83c2263282a9068655e8e3fbbc167b653/docs/faq.md#can-external-dns-manageaddremove-records-in-a-hosted-zone-which-is-setup-in-different-aws-account)

Reviewer 3: I think we can mention the fact that we considered [the external-id flag](https://github.com/kubernetes-sigs/external-dns/blob/22da9f231dbc6faa3a668b507a4a06823a129609/pkg/apis/externaldns/types.go#L460) and decided not to implement for now.

Reviewer 3: I didn't think during the first review but may be `external-id` flag can be considered as an alternative, so placed into `Alternative` section?

Author: Sure, that makes sense. I made it a bit more generic to apply to OCM generally and ExternalDNS and moved it to Alternatives.

argument, which uses the specified AWS role ARN when creating new DNS records. Therefore, to support shared VPCs with
the ExternalDNS Operator, we update the [API](#API-External-DNS-Operator) to expose the `--aws-assume-role` argument
for the ExternalDNS binary.

By design, ExternalDNS can update both public and private DNS zones using credentials that the cluster-admin provides.
Therefore, ExternalDNS Operator users are able to use the new [API](#API-External-DNS-Operator) to update both public
and private DNS zones in other accounts, provided the role ARN has appropriate permissions.

More details on using a role ARN to create DNS records in another AWS account will be documented in ExternalDNS
Operator's [Docs](https://github.com/openshift/external-dns-operator/tree/main/docs).

> 

Reviewer 3: We can mention the update of the existing webhook for the aws provider. As you found, it currently forbids the usage of `provider.aws` field without `credentials`.

Reviewer 3: True, I overlooked this. The webhook was just checking for the emptiness of the credentials secret name.

An e2e test will be setup which utilizes both AWS CI accounts.

For the ExternalDNS Operator, we will add e2e tests utilizing both AWS CI accounts using the RoleARN to create DNS
records.

### Graduation Criteria

This functionality is targeted for 4.14 GA and for backporting to previous releases.

Note that the External-DNS-Operator updates to support Shared VPC are out of payload and are therefore not aligned with 
OpenShift releases and will not need backporting.

> #### Dev Preview -> Tech Preview

Reviewer 3: I suppose the update for ExternalDNS Operator API with the assume role field will not be guarded by a feature gate as it's done for the cluster ingress operator. Probably worth a mention as it may give an impression that the feature gate is still needed.

Author: Done.

preferred as it allows users to declare intent of a cross account install,
rather than OpenShift trying to infer that from the credentials.

### External ID

We considered supporting configuration of an [external ID](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_roles_create_for-user_externalid.html)
for the Ingress Operator when using an assumed role. AWS suggests the use of an external ID in specific scenarios to

> mitigate the confused deputy problem. However, for this effort, we opted not to include this functionality at the

Author: Whoops, skimming over old comments, I realized I lost this change. Done in next push.
